crates/extensions/c8y_mapper_ext/src/operations/mod.rs

```diff
@@ -210,74 +222,123 @@ impl RunningOperation {
-                            .handle_log_upload_state_change(entity, &cmd_id, &message)
+                            .handle_log_upload_state_change(&entity, &cmd_id, &message)
                             .await
                     }
                     OperationType::ConfigSnapshot => {
                         context
-                            .handle_config_snapshot_state_change(entity, &cmd_id, &message)
+                            .handle_config_snapshot_state_change(&entity, &cmd_id, &message)
                             .await
                     }
                     OperationType::ConfigUpdate => {
                         context
-                            .handle_config_update_state_change(entity, &cmd_id, &message)
+                            .handle_config_update_state_change(&entity, &cmd_id, &message)
                             .await
                     }
                     OperationType::FirmwareUpdate => {
                         context
-                            .handle_firmware_update_state_change(entity, &cmd_id, &message)
+                            .handle_firmware_update_state_change(&entity, &cmd_id, &message)
                             .await
                     }
                 };
 
                 let mut mqtt_publisher = context.mqtt_publisher.clone();
-                match res {
-                    // If there are mapped final status messages to be published, they are cached until the operation
-                    // log is uploaded
-                    Ok((messages, command)) => {
-                        if let Some(command) = command {
-                            if let Err(e) = context
-                                .upload_operation_log(&external_id, &cmd_id, &operation, command)
-                                .await
-                            {
-                                error!("failed to upload operation logs: {e}");
-                            }
+
+                // unwrap is safe: at this point all local operations that are not regular c8y
+                // operations should be handled above
+                let c8y_operation = to_c8y_operation(&operation).unwrap();
+
+                match to_response(
+                    operation_result,
+                    c8y_operation,
+                    &entity.smartrest_publish_topic,
+                ) {
+                    OperationResult::Ignored => {}
+                    OperationResult::Executing => {
+                        let c8y_state_executing_payload = set_operation_executing(c8y_operation);
+                        let c8y_state_executing_message = MqttMessage::new(
+                            &entity.smartrest_publish_topic,
+                            c8y_state_executing_payload,
+                        );
+                        mqtt_publisher
+                            .send(c8y_state_executing_message)
+                            .await
+                            .unwrap();
```

Comment:
> And these terminated tasks will be removed from `running_operations` when the next message for that operation type is handled.

Not next message for that operation type, but next message for that particular operation, with that particular id.

But I think there's a problem: in practice, when a panic in a child task happens, `.handle()` will never get called for that operation again. Operation task itself sends the messages, so if they can't be sent over MQTT because of a panic, they won't be handled, won't be joined, so we will not see that an exception has occurred.

For now, in this PR I will add a test to verify the behaviour of the `handle` function - when does it panic and when does it actually join the tasks.

But what I realize is that `OperationHandler` needs its own background task that joins the operation tasks (i.e. polls next future in `FuturesUnordered`), which will have to be added, or I'll have to make `OperationHandler` an actor, which I probably should. So far I've avoided it because I wanted `OperationHandler` to be a small object used and managed by the converter, not another actor, and also because our actors tend to have a lot of boilerplate, but the time for that is probably now and it's what I'll have to do in a follow-up PR.

So AFAIK right now we don't actually join the panicked tasks and don't know that a panic has occurred. This is the first thing that I should fix. Then, I'll need to make sure tasks are joined automatically so `.handle()` doesn't have to be called to know whether or not there's been a panic. I leave to you the decision of whether these changes should be made in a follow-up PR, or in this one.

Reply:
> I'll have to make OperationHandler an actor, which I probably should

Yeah, that's pretty much aligned with the model that I had in mind too, esp after the introduction of the availability actor.

> but the time for that is probably now and it's what I'll have to do in a follow-up PR.

That's fine. One step at a time.